### PR TITLE
Rework the handling of profiles

### LINKED ISSUE #155: Tycho publishes too much
I run a pure J11 build (j11 jdk and all bundles require JavaSE11) Tycho still instals profiles for 14,15,16 ...

```
[INFO] --- tycho-p2-publisher-plugin:2.2.0:publish-osgi-ee (default-publish-osgi-ee) @ de.laeubisoft.products.labratry.sgs ---
[INFO] Published profile IUs: [a.jre.javase 11.0.0, config.a.jre.javase 11.0.0]
[INFO] Published profile IUs: [config.a.jre.javase 14.0.0, a.jre.javase 14.0.0]
[INFO] Published profile IUs: [config.a.jre.javase 15.0.0, a.jre.javase 15.0.0]
[INFO] Published profile IUs: [config.a.jre.javase 16.0.0, a.jre.javase 16.0.0]
```

### LINKED ISSUE #276: Tycho should find Toolchain JRE by version
Currently if there is a toolchain configured this way:

```
    <toolchain>
        <type>jdk</type>
        <provides>
            <version>1.8</version>
        </provides>
        <configuration>
            <jdkHome>${env.JAVA_HOME_8_X64}</jdkHome>
        </configuration>
    </toolchain>
```

Tycho fails to find this jdk with `useJDK = BREE configured, but no toolchain of type 'jdk' with id 'JavaSE-1.8'` it would be good if tycho (when not finding a jdk by id) simply search one of type jdk+version=1.8 in this case

### LINKED ISSUE #655: Remove warning about No system packages found in profile nor toolchain for missing JRE
I use a certain Java release to build, currently Java 17, in my Tycho build I see lots of warnings.

I already set           
`<resolveWithExecutionEnvironmentConstraints>false</resolveWithExecutionEnvironmentConstraints>`

Is their somthing missing on my side to get rid of these warnings? Of does the messae not use this setting?

[WARNING] No system packages found in profile nor toolchain for JRE-1.1, using current JRE system packages.
This can cause faulty dependency resolution, consider adding a definition for a 'jdk' with id=JRE-1.1 in your toolchains.xml
[WARNING] No system packages found in profile nor toolchain for JavaSE-11, using current JRE system packages.
This can cause faulty dependency resolution, consider adding a definition for a 'jdk' with id=JavaSE-11 in your toolchains.xml
[WARNING] No system packages found in profile nor toolchain for JavaSE-16, using current JRE system packages.
This can cause faulty dependency resolution, consider adding a definition for a 'jdk' with id=JavaSE-16 in your toolchains.xml
[WARNING] No system packages found in profile nor toolchain for JavaSE-18, using current JRE system packages.
This can cause faulty dependency resolution, consider adding a definition for a 'jdk' with id=JavaSE-18 in your toolchains.xml

### LINKED ISSUE #958: Unknown OSGi execution environment: 'JavaSE-19'
I'm seeing the following error trying to use tycho with a JDK 19-ea:

```
Error:  Internal error: org.eclipse.tycho.core.ee.UnknownEnvironmentException: Unknown OSGi execution environment: 'JavaSE-19' -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: org.eclipse.tycho.core.ee.UnknownEnvironmentException: Unknown OSGi execution environment: 'JavaSE-19'
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.eclipse.tycho.core.ee.UnknownEnvironmentException: Unknown OSGi execution environment: 'JavaSE-19'
    at org.eclipse.tycho.core.ee.ExecutionEnvironmentUtils.getExecutionEnvironment (ExecutionEnvironmentUtils.java:96)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.applyBestOfCurrentOrConfiguredProfile (OsgiBundleProject.java:682)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.readExecutionEnvironmentConfiguration (OsgiBundleProject.java:[65](https://github.com/google/google-java-format/runs/6392813704?check_suite_focus=true#step:5:65)2)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.setupProject (DefaultTychoResolver.java:108)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:111)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:253)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

### PR DESCRIPTION
**Pull request by laeubi**
- do not embed any static profile
- compute the java packages based on the running vm and toolchains
- fall back to "higher" profiles if an exact match is not found
- do not validate BREEs against static list of names

 
- Fixes #958
- Fixes #655
- Fixes #276
- Fixes #155

### CODE REVIEW

File: tycho-core/src/test/java/org/eclipse/tycho/core/ee/StandardExecutionEnvironmentTest.java [hunk 1]
```diff
@@ -27,9 +27,7 @@
 
 public class StandardExecutionEnvironmentTest {
 
-    private StandardExecutionEnvironment javaSE18Environment;
-    private StandardExecutionEnvironment javaSE17Environment;
-    private StandardExecutionEnvironment javaSE11Environment;
+    private StandardExecutionEnvironment runningJavaEnvironment;
     private StandardExecutionEnvironment javaSE9Environment;
     private StandardExecutionEnvironment javaSE8Environment;
     private StandardExecutionEnvironment javaSE7Enviroment;
```
Comment (Bananeweizen) [inline tycho-core/src/test/java/org/eclipse/tycho/core/ee/StandardExecutionEnvironmentTest.java:33]: you could do me a favor by search/replacing the typo "enviroment" (missing n) :)

Comment (laeubi) [reply tycho-core/src/test/java/org/eclipse/tycho/core/ee/StandardExecutionEnvironmentTest.java]: Done

**Comment by laeubi** [comment]
@akurtakov there is a "Java18ResolutionTest" that currently fails because the highest JDK available to our Jenkins build is java 17 (for github actions) I see the following options:

1. we add java 18 as well to the toolchain of gh-action
2. we simply drop that test unless we plan to upgrade to Java 19
3. we add an ignore to that test

all other tests succeed, don'T know whats wrong with the jenkins CI  ...

**Comment by akurtakov** [comment]
I'm quite busy with 4.24/4.25 stream duties. Once done there I'll be back on tycho, hopefully @mickaelistria beats me to it.

**Comment by akurtakov** [comment]
Adding Java 18 to the toolchain sounds best option to me.

**Comment by laeubi** [comment]
> Adding Java 18 to the toolchain sounds best option to me.

I have added Java 18 now.

**Comment by laeubi** [comment]
Currently the setup-java task produces weird results: https://github.com/actions/setup-java/issues/337 ...

**Comment by laeubi** [comment]
I disabled the Java 18 test for now unless either
- https://github.com/actions/setup-java/issues/276
- https://github.com/actions/setup-java/issues/44

is fixed, @eclipse/technology-tycho-committers would be good to express your support for any or both tickets to be fixed, this will essentially be useful for SWT as well.

**Comment by laeubi** [comment]
I also opened: https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1388

File: tycho-core/src/main/java/org/eclipse/tycho/core/ee/ListSystemPackages.java [hunk 2]
```diff
@@ -19,17 +21,27 @@
...
+                        desc.exports().stream().filter(not(java.lang.module.ModuleDescriptor.Exports::isQualified))
                                 .map(java.lang.module.ModuleDescriptor.Exports::source) //
                 ).collect(Collectors.toSet());
     }
 
+    private static Predicate<? super Exports> not(Predicate<Exports> predicate) {
+        // java 9 compatible polyfill
+        return x -> !predicate.test(x);
+    }
+
```
Comment (mickaelistria) [inline tycho-core/src/main/java/org/eclipse/tycho/core/ee/ListSystemPackages.java:46]: Can you please extract this into a dedicated patch?

**Comment by mickaelistria** [comment]
I like the general idea and it's a natural continuation of other changed that have happened on this topic over last versions.

I believe there are some parts of this patch that could be isolated a bit as they are not really contributing directly to the goal and are more preliminary fix/cleanups. It would be nice if you can separate separate commits for those; that would greatly facilitate review and also history navigation when we need to look back at this change in the future..

File: tycho-core/src/main/java/org/eclipse/tycho/core/ee/ListSystemPackages.java (hunk 2 above)
Comment (laeubi) [reply tycho-core/src/main/java/org/eclipse/tycho/core/ee/ListSystemPackages.java:46]: https://github.com/eclipse/tycho/pull/996

File: tycho-core/src/main/java/org/eclipse/tycho/core/ee/StandardExecutionEnvironment.java [hunk 3]
```diff
@@ -110,29 +110,56 @@ public class StandardExecutionEnvironment implements Comparable<StandardExecutio
...
+                    }
+                } catch (NumberFormatException e) {
+                    StringBuilder sb = new StringBuilder(line);
+                    while ((line = reader.readLine()) != null) {
+                        sb.append(System.lineSeparator());
+                        sb.append(line);
+                    }
+                    logger.debug("[ReadPackagesFromToolchains] Can't read java version for " + java
+                            + ", full output was: " + sb);
+                    return new JavaInfo(-1, List.of());
```
Comment (mickaelistria) [inline tycho-core/src/main/java/org/eclipse/tycho/core/ee/StandardExecutionEnvironment.java:142]: Wouldn't it be more explicit to just return nothing here instead of returning an incorrect object?

File: .github/workflows/maven.yml [hunk 4]
```diff
@@ -20,6 +20,11 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v3
+    - name: Set up JDK 18
+      uses: actions/setup-java@v3
+      with:
+        java-version: '18'
+        distribution: 'temurin'
```
[the remark this answers is not in the document]
Comment (laeubi) [reply .github/workflows/maven.yml:27]: I'll simply revert this as it does not work as intended we would need 

- https://github.com/actions/setup-java/issues/44

and

- https://github.com/actions/setup-java/issues/276

for Java > 17 support in GH actions, so please express your interest in getting these fixes (where one already has a fixing PR).

**Comment by laeubi** [comment]
Rebased and remove unnecessary changes... @mickaelistria anything you think should be extracted or are we fine here?

**Comment by mickaelistria** [comment]
Thanks for this great change, it allows to support newer Java versions without need to change anything in Tycho. Awesome!
